File: src/procedures/FillStripProcedure.java
package procedures;

import java.awt.Color;

import led.LEDDataBundle;
import led.LEDStripManager;
import procedures.models.Direction;

//Animation to fill the strip from a given direction (left, right, center, bounds) with a given color 
public class FillStripProcedure extends Procedure {
	private Color mFillColor = Color.BLACK;
	
	public int mSpeed = 1;
	public Direction mDirection = Direction.Left;
	
	public FillStripProcedure(LEDDataBundle _bundle) {
		super(_bundle);
		
		mFillColor = _bundle.colorPrimary.toSystemColor();
		mSpeed = Math.round(_bundle.speed);
		mDirection = _bundle.direction;
		
		mSteps = LEDStripManager.LED_COUNT;
		
		if (mDirection == Direction.Center || mDirection == Direction.CenterInvert) {
			mSteps /= 2;
		}
	}
	
	@Override
	public void start() {
		
	}
	
	@Override
	void update() {
		switch (mDirection) {
		case Center:
			for (int i = 0; i < mSpeed; i++) {
				mStrip.setPixel((LEDStripManager.LED_COUNT / 2 - 1) + mStep + i, mFillColor);
				mStrip.setPixel((LEDStripManager.LED_COUNT / 2 - 1) - mStep - i, mFillColor);	
			}
			break;
		case CenterInvert:
			for (int i = 0; i < mSpeed; i++) {
				mStrip.setPixel(mStep + i, mFillColor);
				mStrip.setPixel((LEDStripManager.LED_COUNT - 1) - mStep - i, mFillColor);	
			}
			break;
		case Left:
			for (int i = 0; i < mSpeed; i++) {
				mStrip.setPixel(mStep + i, mFillColor);	
			}
			break;
		case Right:
			for (int i = 0; i < mSpeed; i++) {
				mStrip.setPixel((LEDStripManager.LED_COUNT - 1) - mStep - i, mFillColor);	
			}
			break;
		default:
			break;
			
		}
		
		mStep += mSpeed;
		
		if (mStep >= mSteps) {
			if (!mIsSubProcedure) {
				finishProcedure();				
			}
		}
	}
}

/*
 * Ideas:
 * - Ease interpolation function
 * - Tiling (single fill as normal, double fill with e.g. "two center points" to fill from
 */

File: src/main/java/procedures/FillStripProcedure.kt
package procedures

import com.github.iamniklas.liocore.led.LEDDataBundle
import com.github.iamniklas.liocore.led.LEDStripManager
import com.github.iamniklas.liocore.procedures.Procedure
import com.github.iamniklas.liocore.procedures.models.Direction
import java.awt.Color

//Animation to fill the strip from a given direction (left, right, center, bounds) with a given color 
class FillStripProcedure(_bundle: LEDDataBundle) : Procedure(_bundle) {
    private var mFillColor = Color.BLACK
    var mSpeed = 1
    var mDirection: Direction? = Direction.Left
    override fun start() {}
    public override fun update() {
        when (mDirection) {
            Direction.Center -> {
                var i = 0
                while (i < mSpeed) {
                    strip!!.setPixel(LEDStripManager.LED_COUNT / 2 - 1 + step + i, mFillColor)
                    strip!!.setPixel(LEDStripManager.LED_COUNT / 2 - 1 - step - i, mFillColor)
                    i++
                }
            }
            Direction.CenterInvert -> {
                var i = 0
                while (i < mSpeed) {
                    strip!!.setPixel(step + i, mFillColor)
                    strip!!.setPixel(LEDStripManager.LED_COUNT - 1 - step - i, mFillColor)
                    i++
                }
            }
            Direction.Left -> {
                var i = 0
                while (i < mSpeed) {
                    strip!!.setPixel(step + i, mFillColor)
                    i++
                }
            }
            Direction.Right -> {
                var i = 0
                while (i < mSpeed) {
                    strip!!.setPixel(LEDStripManager.LED_COUNT - 1 - step - i, mFillColor)
                    i++
                }
            }
            else -> {
            }
        }
        step += mSpeed
        if (step >= steps) {
            if (!isSubProcedure) {
                finishProcedure()
            }
        }
    }

    init {
        mFillColor = _bundle.colorPrimary!!.toSystemColor()
        mSpeed = Math.round(_bundle.speed!!)
        mDirection = _bundle.direction
        steps = LEDStripManager.LED_COUNT
        if (mDirection == Direction.Center || mDirection == Direction.CenterInvert) {
            steps /= 2
        }
    }
}